thingsboard_gateway/connectors/modbus_async/modbus_connector.py

```diff
@@ -0,0 +1,697 @@
+                self.__log.exception('Failed to add slave: %s', e)
+
+        self.__log.debug('Added %d slaves', len(self.__slaves))
+
+    @classmethod
+    def callback(cls, slave: Slave):
+        cls.PROCESS_REQUESTS.put_nowait(slave)
+
+    async def __process_requests(self):
+        while not self.__stopped:
+            try:
+                slave = self.PROCESS_REQUESTS.get_nowait()
+                await self.__poll_device(slave)
+            except Empty:
+                await asyncio.sleep(.01)
+            except Exception as e:
+                self.__log.exception(e)
+
+    async def __poll_device(self, slave):
+        self.__log.debug("Polling %s slave", slave)
+
+        if slave.type == 'serial':
+            await self.__lock.acquire()
+
+        # check if device have attributes or telemetry to poll
+        if slave.uplink_converter_config.attributes or slave.uplink_converter_config.telemetry:
+            try:
+                connected_to_master = await slave.connect()
+
+                self.__manage_device_connectivity_to_platform(slave)
+
+                if connected_to_master:
+                    slave_data = await self.__read_slave_data(slave)
+                    self.__data_to_convert.put_nowait((slave, slave_data))
+                else:
+                    self.__log.error('Socket is closed, connection is lost, for device %s', slave)
+            except (ConnectionException, asyncio.exceptions.TimeoutError):
+                self.__delete_device_from_platform(slave)
+                await asyncio.sleep(5)
+                self.__log.error('Failed to connect to device %s', slave)
+            except Exception as e:
+                self.__delete_device_from_platform(slave)
+                self.__log.exception(e)
+            finally:
+                if slave.type == 'serial':
+                    self.__lock.release()
+        else:
+            self.__log.error('Config is empty. Nothing to read, for device %s', slave)
+
+    async def __read_slave_data(self, slave: Slave):
+        result = {
+            'telemetry': {},
+            'attributes': {}
+        }
+
+        for config_section in ('attributes', 'telemetry'):
+            for config in getattr(slave.uplink_converter_config, config_section):
+                response = await slave.read(config['functionCode'], config['address'], config['objectsCount'])
+
+                if 'Exception' in str(response) or 'Error' in str(response):
+                    self.__log.error("Reading failed for device %s function code %s address %s unit id %s",
+                                     slave.device_name, config['functionCode'], config[ADDRESS_PARAMETER],
+                                     slave.unit_id)
+                    self.__log.error("Reading failed with exception:", exc_info=result)
+                    self.__log.info("Trying to reconnect to device %s", slave.device_name)
+                    if slave.master.connected:
```

Comment:
It is preferable to keep connection to master opened and not close it after each reading/writing operation.